client/service/custom_server_manager.go

```
@@ -0,0 +1,312 @@
+package boxservice
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/sagernet/sing-box/adapter"
+	"github.com/sagernet/sing-box/constant"
+	"github.com/sagernet/sing-box/log"
+	"github.com/sagernet/sing-box/option"
+	"github.com/sagernet/sing-box/protocol/group"
+	"github.com/sagernet/sing/common/json"
+	"github.com/sagernet/sing/service"
+)
+
+type CustomServerManager struct {
+	ctx                   context.Context
+	customServersMutex    *sync.RWMutex
+	customServers         map[string]CustomServerInfo
+	customServersFilePath string
+}
+
+func NewCustomServerManager(ctx context.Context, dataDir string) *CustomServerManager {
+	return &CustomServerManager{
+		ctx:                   ctx,
+		customServers:         make(map[string]CustomServerInfo),
+		customServersMutex:    new(sync.RWMutex),
+		customServersFilePath: filepath.Join(dataDir, "data", "custom_servers.json"),
+	}
+}
+
+type customServers struct {
+	CustomServers []CustomServerInfo `json:"custom_servers"`
+}
+
+// CustomServerInfo represents a custom server configuration.
+type CustomServerInfo struct {
+	Tag      string           `json:"tag"`
+	Outbound *option.Outbound `json:"outbound,omitempty"`
+	Endpoint *option.Endpoint `json:"endpoint,omitempty"`
+}
+
+// ServerConnectConfig represents configuration for connecting to a custom server.
+type ServerConnectConfig []byte
+
+// AddCustomServer load or parse the given configuration and add given
+// endpdoint/outbound to the instance. We're only expecting one endpoint or
+// outbound per call.
+func (m *CustomServerManager) AddCustomServer(tag string, cfg ServerConnectConfig) error {
+	m.customServersMutex.Lock()
+	loadedOptions := m.customServers[tag]
+	m.customServersMutex.Unlock()
+
+	if cfg != nil {
```

There's a bug here; we can't add a server if `cfg` is `nil`, so we should return an error if it is.

---

I've added a verification for checking if the tag, endpoint or outbound are empty or nil